Build fails for me, this is the only line in red so assume the related on.

src/engine/cachingreader.cpp:369:34: error: ‘bufferedFrameIndexRangee’ was not declared in this scope
                     DEBUG_ASSERT(bufferedFrameIndexRangee.empty());

--- a/src/engine/cachingreader.cpp
+++ b/src/engine/cachingreader.cpp
@@ -220,9 +220,21 @@ void CachingReader::process() {
     }
 }
 
-SINT CachingReader::read(SINT startSample, SINT numSamples, bool reverse, CSAMPLE* buffer) {
+bool CachingReader::read(SINT startSample, SINT numSamples, bool reverse, CSAMPLE* buffer) {
+    // If asked to read 0 samples, don't do anything. (this is a perfectly
+    // reasonable request that happens sometimes.
+    if (numSamples == 0) {
+        return true;
+    }
+
     VERIFY_OR_DEBUG_ASSERT(buffer) {
-        return 0;
+        return false;
+    }
+
+    // If no track is loaded, don't do anything.
+    if (m_readerStatus != TRACK_LOADED) {
+        SampleUtil::clear(buffer, numSamples);
+        return false;
     }
 
     // the samples are always read in forward direction
@@ -247,13 +259,6 @@ SINT CachingReader::read(SINT startSample, SINT numSamples, bool reverse, CSAMPL
         QString temp = QString("Sample = %1").arg(sample);
         kLogger.debug() << "read() invalid arguments sample:" << sample
                  << "numSamples:" << numSamples << "buffer:" << buffer;
-        return 0;
-    }
-
-    // If asked to read 0 samples, don't do anything. (this is a perfectly
-    // reasonable request that happens sometimes. If no track is loaded, don't
-    // do anything.
-    if (numSamples == 0 || m_readerStatus != TRACK_LOADED) {
         return 0;
     }
 
@@ -369,6 +374,12 @@ SINT CachingReader::read(SINT startSample, SINT numSamples, bool reverse, CSAMPL
                     DEBUG_ASSERT(bufferedFrameIndexRange.empty());
                 }
                 if (bufferedFrameIndexRange.empty()) {
+                    if (chunkIndex == firstChunkIndex) {
+                        // We have not read a single frame.
+                        // Inform the calling code by returning 0
+                        // that it can fade in the next time
+                        return false;
+                    }
                     // No more readable data available. Exit the loop and
                     // finally fill the remaining buffer with silence.
                     break;
@@ -409,7 +420,7 @@ SINT CachingReader::read(SINT startSample, SINT numSamples, bool reverse, CSAMPL
     // Finally fill the remaining buffer with silence
     DEBUG_ASSERT(samplesRemaining >= 0);
     SampleUtil::clear(buffer, samplesRemaining);
-    return numSamples;
+    return true;
 }
 
 void CachingReader::hintAndMaybeWake(const HintVector& hintList) {

--- a/src/engine/cachingreader.cpp
+++ b/src/engine/cachingreader.cpp
@@ -343,39 +343,34 @@ SINT CachingReader::read(SINT startSample, SINT numSamples, bool reverse, CSAMPL
                     break;
                 }
 
+                mixxx::IndexRange bufferedFrameIndexRange;
                 const CachingReaderChunkForOwner* const pChunk = lookupChunkAndFreshen(chunkIndex);
-                // If the chunk is not in cache, then we must return an error.
-                if (!pChunk || (pChunk->getState() != CachingReaderChunkForOwner::READY)) {
+                if (pChunk && (pChunk->getState() == CachingReaderChunkForOwner::READY)) {
+                    if (reverse) {
+                        bufferedFrameIndexRange =
+                                pChunk->readBufferedSampleFramesReverse(
+                                        &buffer[samplesRemaining],
+                                        remainingFrameIndexRange);
+                    } else {
+                        bufferedFrameIndexRange =
+                                pChunk->readBufferedSampleFrames(
+                                        buffer,
+                                        remainingFrameIndexRange);
+                    }
+                } else {
                     // This will happen regularly when jumping to a new position
                     // within the file and decoding of the audio data is still
                     // pending.
                     DEBUG_ASSERT(!pChunk ||
                             (pChunk->getState() == CachingReaderChunkForOwner::READ_PENDING));
                     Counter("CachingReader::read(): Failed to read chunk on cache miss")++;
-                    // Abort reading and indicate (temporary) unavailability
-                    // by returning 0. Some samples may have already been
-                    // written into the caller's buffer, but should not be
-                    // used. The caller should retry if it still needs to
-                    // read those samples.
-                    kLogger.debug() << "Cache miss for chunk with index" << chunkIndex;
-                    return 0;
-                }
-
-                mixxx::IndexRange bufferedFrameIndexRange;
-                if (reverse) {
-                    bufferedFrameIndexRange =
-                            pChunk->readBufferedSampleFramesReverse(
-                                    &buffer[samplesRemaining],
-                                    remainingFrameIndexRange);
-                } else {
-                    bufferedFrameIndexRange =
-                            pChunk->readBufferedSampleFrames(
-                                    buffer,
-                                    remainingFrameIndexRange);
+                    kLogger.debug() << "Cache miss for chunk with index" << chunkIndex << "- abort reading";
+                    // Abort reading (see below)
+                    DEBUG_ASSERT(bufferedFrameIndexRange.empty());
                 }
                 if (bufferedFrameIndexRange.empty()) {
                     // No more readable data available. Exit the loop and
-                    // fill the remaining buffer with silence.
+                    // finally fill the remaining buffer with silence.
                     break;
                 }
                 DEBUG_ASSERT(bufferedFrameIndexRange <= remainingFrameIndexRange);
@@ -401,6 +396,7 @@ SINT CachingReader::read(SINT startSample, SINT numSamples, bool reverse, CSAMPL
                 }
                 const SINT chunkSamples =
                         CachingReaderChunk::frames2samples(bufferedFrameIndexRange.length());
+                DEBUG_ASSERT(chunkSamples > 0);
                 if (!reverse) {
                     buffer += chunkSamples;
                 }

--- a/src/engine/cachingreader.h
+++ b/src/engine/cachingreader.h
@@ -83,9 +83,10 @@ class CachingReader : public QObject {
     virtual void process();
 
     // Read numSamples from the SoundSource starting with sample into
-    // buffer. Returns the total number of samples actually written to buffer
-    // support reading stereo samples in reverse (backward) order
-    virtual SINT read(SINT startSample, SINT numSamples, bool reverse, CSAMPLE* buffer);
+    // buffer. It always writes numSamples to the buffer. In case of a chach miss,
+    // it returns false and clears the buffer.
+    // It support reading stereo samples in reverse (backward) order
+    virtual bool read(SINT startSample, SINT numSamples, bool reverse, CSAMPLE* buffer);
 
     // Issue a list of hints, but check whether any of the hints request a chunk
     // that is not in the cache. If any hints do request a chunk not in cache,

--- a/src/engine/enginebuffer.cpp
+++ b/src/engine/enginebuffer.cpp
@@ -983,6 +983,7 @@ void EngineBuffer::process(CSAMPLE* pOutput, const int iBufferSize) {
             // Perform scaling of Reader buffer into buffer.
             double framesRead =
                     m_pScale->scaleBuffer(pOutput, iBufferSize);
+
             // TODO(XXX): The result framesRead might not be an integer value.
             // Converting to samples here does not make sense. All positional
             // calculations should be done in frames instead of samples! Otherwise

--- a/src/engine/readaheadmanager.cpp
+++ b/src/engine/readaheadmanager.cpp
@@ -17,7 +17,8 @@ ReadAheadManager::ReadAheadManager()
           m_pRateControl(NULL),
           m_currentPosition(0),
           m_pReader(NULL),
-          m_pCrossFadeBuffer(SampleUtil::alloc(MAX_BUFFER_LEN)) {
+          m_pCrossFadeBuffer(SampleUtil::alloc(MAX_BUFFER_LEN)),
+          m_cacheMissHappened(false) {
     // For testing only: ReadAheadManagerMock
 }
 
@@ -27,7 +28,8 @@ ReadAheadManager::ReadAheadManager(CachingReader* pReader,
           m_pRateControl(NULL),
           m_currentPosition(0),
           m_pReader(pReader),
-          m_pCrossFadeBuffer(SampleUtil::alloc(MAX_BUFFER_LEN)) {
+          m_pCrossFadeBuffer(SampleUtil::alloc(MAX_BUFFER_LEN)),
+          m_cacheMissHappened(false) {
     DEBUG_ASSERT(m_pLoopingControl != NULL);
     DEBUG_ASSERT(m_pReader != NULL);
 }
@@ -90,7 +92,12 @@ SINT ReadAheadManager::getNextSamples(double dRate, CSAMPLE* pOutput,
     // in case of a chache miss the buffer is cleared and the return value is false;
 
     if (!readSuccess) {
-        qDebug() << "##################################";
+        m_cacheMissHappened = true;
+    } else if (m_cacheMissHappened) {
+        m_cacheMissHappened = false;
+        // Apply raming gain, because the last buffer has unwanted silenced and
+        // new without fading are causing a pop.
+        SampleUtil::applyRampingGain(pOutput, 0.0, 1.0, samples_from_reader);
     }
 
     // Increment or decrement current read-ahead position
@@ -138,6 +145,7 @@ SINT ReadAheadManager::getNextSamples(double dRate, CSAMPLE* pOutput,
 
         if (!readSuccess2) {
             qDebug() << "ERROR: Couldn't get all needed samples for crossfade.";
+            m_cacheMissHappened = true;
         }
 
         // do crossfade from the current buffer into the new loop beginning
@@ -157,6 +165,7 @@ void ReadAheadManager::addRateControl(RateControl* pRateControl) {
 // Not thread-save, call from engine thread only
 void ReadAheadManager::notifySeek(double seekPosition) {
     m_currentPosition = seekPosition;
+    m_cacheMissHappened = false;
     m_readAheadLog.clear();
 
     // TODO(XXX) notifySeek on the engine controls. EngineBuffer currently does

--- a/src/engine/readaheadmanager.cpp
+++ b/src/engine/readaheadmanager.cpp
@@ -84,22 +84,24 @@ SINT ReadAheadManager::getNextSamples(double dRate, CSAMPLE* pOutput,
     SINT start_sample = SampleUtil::roundPlayPosToFrameStart(
             m_currentPosition, kNumChannels);
 
-    SINT samples_read = m_pReader->read(
+    bool readSuccess = m_pReader->read(
             start_sample, samples_from_reader, in_reverse, pOutput);
+    // read() returns a buffer filed up to "samples_from_reader".
+    // in case of a chache miss the buffer is cleared and the return value is false;
 
-    if (samples_read != samples_from_reader) {
-        qDebug() << "didn't get what we wanted" << samples_read << samples_from_reader;
+    if (!readSuccess) {
+        qDebug() << "##################################";
     }
 
     // Increment or decrement current read-ahead position
     // Mixing int and double here is desired, because the fractional frame should
     // be resist
     if (in_reverse) {
-        addReadLogEntry(m_currentPosition, m_currentPosition - samples_read);
-        m_currentPosition -= samples_read;
+        addReadLogEntry(m_currentPosition, m_currentPosition - samples_from_reader);
+        m_currentPosition -= samples_from_reader;
     } else {
-        addReadLogEntry(m_currentPosition, m_currentPosition + samples_read);
-        m_currentPosition += samples_read;
+        addReadLogEntry(m_currentPosition, m_currentPosition + samples_from_reader);
+        m_currentPosition += samples_from_reader;
     }
 
     // Activate on this trigger if necessary
@@ -131,21 +133,21 @@ SINT ReadAheadManager::getNextSamples(double dRate, CSAMPLE* pOutput,
         int loop_read_position = SampleUtil::roundPlayPosToFrameStart(
                 m_currentPosition + (in_reverse ? preloop_samples : -preloop_samples), kNumChannels);
 
-        int looping_samples_read = m_pReader->read(
-                loop_read_position, samples_read, in_reverse, m_pCrossFadeBuffer);
+        bool readSuccess2 = m_pReader->read(
+                loop_read_position, samples_from_reader, in_reverse, m_pCrossFadeBuffer);
 
-        if (looping_samples_read != samples_read) {
+        if (!readSuccess2) {
             qDebug() << "ERROR: Couldn't get all needed samples for crossfade.";
         }
 
         // do crossfade from the current buffer into the new loop beginning
-        if (samples_read != 0) { // avoid division by zero
-            SampleUtil::linearCrossfadeBuffers(pOutput, pOutput, m_pCrossFadeBuffer, samples_read);
+        if (samples_from_reader != 0) { // avoid division by zero
+            SampleUtil::linearCrossfadeBuffers(pOutput, pOutput, m_pCrossFadeBuffer, samples_from_reader);
         }
     }
 
     //qDebug() << "read" << m_currentPosition << samples_read;
-    return samples_read;
+    return samples_from_reader;
 }
 
 void ReadAheadManager::addRateControl(RateControl* pRateControl) {

--- a/src/engine/readaheadmanager.h
+++ b/src/engine/readaheadmanager.h
@@ -125,6 +125,7 @@ class ReadAheadManager {
     double m_currentPosition;
     CachingReader* m_pReader;
     CSAMPLE* m_pCrossFadeBuffer;
+    bool m_cacheMissHappened;
 };
 
 #endif // READAHEADMANGER_H

--- a/src/test/readaheadmanager_test.cpp
+++ b/src/test/readaheadmanager_test.cpp
@@ -17,12 +17,12 @@ class StubReader : public CachingReader {
     StubReader()
             : CachingReader("[test]", UserSettingsPointer()) { }
 
-    SINT read(SINT startSample, SINT numSamples, bool reverse,
+    bool read(SINT startSample, SINT numSamples, bool reverse,
              CSAMPLE* buffer) override {
         Q_UNUSED(startSample);
         Q_UNUSED(reverse);
         SampleUtil::clear(buffer, numSamples);
-        return numSamples;
+        return true;
     }
 };
 

